### DISCUSSION TITLE
Differences between compat and non-compat build

### DISCUSSION THREAD

**Discussion by Spatzendame** [zlib-ng/zlib-ng#966] [General]
Hi,
the question is in the title. Is zlib-ng (compat) more focused on compatibility to original zlib1, and zlib-ng (non-compat) focused on new features?
So which version should I use to replace an original build of zlib1.dll?

Thank you for answer.

**Comment by mtl1979** [comment]
The compatibility version is mostly useful when you can't rebuild the binary using zlib. For anything that has source code available, one should use non-compatibility version and adapt the source codes instead.

Non-compatibility mode fixes several issues due to old code in zlib being designed for 32-bit systems and older even though modern systems are mostly 64-bit and can use larger files and more memory.

**Reply by Spatzendame** [reply]
Okay, so when I have a video game that uses an old zlib1.dll in its installation directory can I replace it with zlib-ng non-compt without to fear about possible conflicts?

**Reply by mtl1979** [reply]
I haven't tried replacing the library under Windows in a while, but last time I checked, it required matching configure options, otherwise Windows reported about missing symbols.

**Comment by nmoinvaz** [comment]
Please see [PORTING.md](https://github.com/zlib-ng/zlib-ng/blob/develop/PORTING.md) for the differences.

**Reply by Spatzendame** [reply]
Thank you both. So without re-compiling an application it is not recommended to use zlib-ng at all? I replaced the zlib1.dll from the game by the zlib-ng and it works. Have the feeling the loading times are a bit faster, but maybe some conflicts can emerge later. So I will stay with the original zlib I think.

The game uses zlib1.dll 1.2.1 (from 2003) so can I just replace it by the latest official zlib 1.2.11 or has the application always to be recompiled?

I noticed applications use different zlib1.dlls (same version number but different in size).

**Reply by mtl1979** [reply]
Like I said earlier, even with different versions of zlib, the configure options must be compatible... You can use version with more features to replace version with less features, but both need to be compiled for either 32-bit or 64-bit processors.

**Reply by Spatzendame** [reply]
Nobody is using a 32-bit processor nowadays, but most software/games are still using 32-bit binaries even on 64-bit Windows and hardware. Do you mean the zlib has to be compiled or the application? I already have zlib compiled as a .dll from its source.

**Reply by mtl1979** [reply]
Both zlib and the application have to be either 32-bit or 64-bit, as those are incompatible.

**Comment by Spatzendame** [comment]
Yes, thank you. I am using 32-bit zlib even on a 64-bit OS and CPU because
the application is 32-bit.

Am Sonntag, 16. Mai 2021 schrieb Mika Lindqvist ***@***.***>:

**Comment by Dead2** [comment]
Replacing a zlib library file (.dll or .so) with the file from zlib-ng compiled in compat mode _can_ work, and probably _usually_ will, but it is not guaranteed and not something we can recommend as it does carry some risk.
For a game, you can easily test and see whether it works or not. If it does not work, it should fail fairly quickly. If you care about your progress in the game in question, make a backup of your save-game files first (google for save-game locations). Even though I think it is not likely to cause data loss in a game, it is possible.

Doing the same for an application that contains vital data, is something we need to warn against, as a crash there could corrupt data that can not easily be replaced. For example on databases, accounting software, medical software, etc, you should be much more careful.

To ensure the least chance of anything going wrong, the application should be recompiled with zlib-ng (in either compat or native mode).
